refactor: replace char_u variables and functions with char

dundargoc opened the pull request:
Work on https://github.com/neovim/neovim/issues/459

dundargoc committed:
refactor: replace char_u variables and functions with char

Work on https://github.com/neovim/neovim/issues/459

dundargoc commented:
Can also replace `vim_strsave` with `xstrdup` and add a `SKIPWHITE` macro at top of file to minimize the number of casts if wanted.

dundargoc commented:
Skipped `get_menu_name` and `get_menu_names` for now as I couldn't quite figure out what was happening with them.

bfredl commented on src/nvim/menu.c:
```diff
@@ -1294,34 +1294,34 @@ static char_u *popup_mode_name(char *name, int idx)
+    text = (char *)vim_strnsave((char_u *)str, (size_t)(p - str));
   } else {
-    text = vim_strsave(str);
+    text = (char *)vim_strsave((char_u *)str);
```
this should just be `xstrdup`

dundargoc replied:
Yeah, figured. I'll fix the other vim_strsave as well :+1:

bfredl commented on src/nvim/menu.c:
```diff
@@ -1294,34 +1294,34 @@ static char_u *popup_mode_name(char *name, int idx)
     }
     assert(p >= str);
-    text = vim_strnsave(str, (size_t)(p - str));
+    text = (char *)vim_strnsave((char_u *)str, (size_t)(p - str));
```
I think this is equivalent to `xstrndup` as well. we should make sure this, because ` (char *)vim_strnsave((char_u *)...)` linenoise will make no-one happy.

dundargoc replied:
Yeah :/ . I'll try to minimize increasing the number of casts as much as possible. Let me know if we should go with another strategy instead. I'm open to ideas :)

Interestingly enough, while `vim_strsave` essentially acts like a wrapper for `xstrdup`, `vim_strnsave` returns `(char_u *)strncpy(xmallocz(len), (char *)string, len)` with the following comment:

>  // strncpy is intentional: some parts of Vim use `string` shorter than `len`
   // and expect the remainder to be zeroed out.

So I assume `vim_strnsave` can be replaced with `xstrndup`, *except* for these "parts", which is a bit vague.

dundargoc replied:
We can otherwise slap a

```c
#define vim_strnsave(a,b) (char*)vim_strnsave((char_u*)a,b)
```

at the top of the file for now and remove it when the time is right?

dundargoc replied:
fixed

bfredl replied:
so they are not the same, as `vim_strnsave` guarantees that `len+1` bytes are allocated even if the valid data is shorter which `xstrndup` doesn't. ideally the latter should work but to be safe rather than sorry we might want a variant that preserves `vim_strnsave` behavior, so we can move forward with the delet char_u project without worry.

dundargoc committed:
refactor: add function xstrnsave

xstrnsave is a clone of vim_strnsave that uses char* instead of char_u*.
Its purpose short-term is to help reduce the number of casts and for
long-term to replace vim_strnsave as the need to use char_u is
eliminated.

justinmk commented:
Dunno if it's related to this, but coverity is now interested in `ex_menu`:

```
*** CID 331356:  Memory - corruptions  (OVERRUN)
/src/nvim/menu.c: 227 in ex_menu()
221               xfree(p);
222             }
223           }
224         }
225     
226         // Careful: remove_menu() changes menu_path
>>>     CID 331356:  Memory - corruptions  (OVERRUN)
>>>     Overrunning buffer pointed to by "menu_path" of 1 bytes by passing it to a function which accesses it at byte offset 1.
227         remove_menu(root_menu_ptr, menu_path, modes, false);
228       } else {
229         /*
230          * Add menu(s).
231          * Replace special key codes.
232          */
```

dundargoc commented:
Coverity says that particular bug was first detected 5th may 2021.
